File: backend/src/scrumix/api/schemas/backlog.py
"""
Backlog-related Pydantic schemas
"""
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field
from scrumix.api.models.backlog import Backlog, BacklogStatus, BacklogPriority, BacklogType
from scrumix.api.schemas.acceptance_criteria import AcceptanceCriteriaResponse


class BacklogBase(BaseModel):
    """Backlog item base information"""
    title: str
    description: Optional[str] = None
    status: BacklogStatus = BacklogStatus.TODO
    story_point: Optional[int] = Field(default=None, ge=0, description="Story points for estimation (must be non-negative)")
    priority: BacklogPriority = BacklogPriority.MEDIUM
    label: Optional[str] = None
    item_type: BacklogType = BacklogType.STORY
    parent_id: Optional[int] = Field(default=None, description="Parent backlog item ID")
    assigned_to_id: Optional[int] = Field(default=None, description="Assigned user ID")


class BacklogCreate(BacklogBase):
    """Create backlog item schema"""
    model_config = ConfigDict(populate_by_name=True)
    
    project_id: int


class BacklogUpdate(BaseModel):
    """Update backlog item schema"""
    model_config = ConfigDict(populate_by_name=True)
    
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[BacklogStatus] = None
    story_point: Optional[int] = Field(default=None, ge=0, description="Story points for estimation (must be non-negative)")
    priority: Optional[BacklogPriority] = None
    label: Optional[str] = None
    item_type: Optional[BacklogType] = None
    parent_id: Optional[int] = Field(default=None, description="Parent backlog item ID")
    assigned_to_id: Optional[int] = Field(default=None, description="Assigned user ID")


class BacklogInDB(BacklogBase):
    """Backlog item information in database"""
    model_config = ConfigDict(from_attributes=True)
    
    backlog_id: int
    created_at: datetime
    updated_at: datetime


class BacklogResponse(BaseModel):
    """Backlog item response schema for frontend"""
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)
    
    id: int
    title: str
    description: Optional[str] = None
    status: BacklogStatus
    story_point: Optional[int] = None
    priority: BacklogPriority
    label: Optional[str] = None
    item_type: BacklogType
    parent_id: Optional[int] = None
    assigned_to_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    # Hierarchical data
    children: Optional[List["BacklogResponse"]] = None
    parent_title: Optional[str] = None  # Title of parent item for display
    
    # Acceptance criteria
    acceptance_criteria: Optional[List["AcceptanceCriteriaResponse"]] = None
    
    @classmethod
    def from_db_model(cls, backlog: "Backlog", children: Optional[List["Backlog"]] = None,
                     parent_title: Optional[str] = None) -> "BacklogResponse":
        """Create response object from database model"""
        children_responses = []
        if children:
            children_responses = [cls.from_db_model(child) for child in children]
        
        # Convert acceptance criteria to response format
        acceptance_criteria_responses = []
        if hasattr(backlog, 'acceptance_criteria') and backlog.acceptance_criteria:
            acceptance_criteria_responses = [
                AcceptanceCriteriaResponse(
                    id=criteria.id,
                    backlog_id=criteria.backlog_id,
                    title=criteria.title,
                    is_met=criteria.is_met,
                    created_at=criteria.created_at,
                    updated_at=criteria.updated_at
                )
                for criteria in backlog.acceptance_criteria
            ]
        
        return cls(
            id=backlog.id,
            title=backlog.title or "",
            description=backlog.description,
            status=backlog.status,
            story_point=backlog.story_point,
            priority=backlog.priority,
            label=backlog.label,
            item_type=backlog.item_type,
            parent_id=backlog.parent_id,
            assigned_to_id=backlog.assigned_to_id,
            created_at=backlog.created_at,
            updated_at=backlog.updated_at,
            children=children_responses if children_responses else None,
            parent_title=parent_title,
            acceptance_criteria=acceptance_criteria_responses if acceptance_criteria_responses else None
        )


# Update forward reference for recursive type
BacklogResponse.model_rebuild() 

File: frontend/src/lib/tools/schemas/backlog.ts
import { z } from 'zod';
import { BacklogStatus, BacklogPriority, BacklogType } from '@/types/api';

/**
 * Schema for creating a backlog item
 * Validates input data before sending to the API - matches monolithic schema exactly
 */
export const createBacklogItemSchema = z.object({
  title: z.string()
    .min(1, 'Title is required')
    .max(200, 'Title must be 200 characters or less')
    .describe('The title of the backlog item'),
  
  description: z.string()
    .max(2000, 'Description must be 2000 characters or less')
    .optional()
    .describe('Optional detailed description of the backlog item'),
  
  priority: z.nativeEnum(BacklogPriority)
    .default(BacklogPriority.MEDIUM)
    .describe('Priority level: critical, high, medium, or low'),
  
  status: z.nativeEnum(BacklogStatus)
    .default(BacklogStatus.TODO)
    .describe('Current status: todo, in_progress, in_review, done, or cancelled'),
  
  item_type: z.nativeEnum(BacklogType)
    .default(BacklogType.STORY)
    .describe('Type of backlog item: epic, story, or bug'),
  
  story_point: z.number()
    .int()
    .min(0, 'Story points must be non-negative')
    .max(100, 'Story points must be 100 or less')
    .optional()
    .describe('Estimation in story points (optional)'),
  
  project_id: z.number()
    .int()
    .positive('Project ID must be a positive integer')
    .describe('The ID of the project this backlog item belongs to'),
  
  parent_id: z.number()
    .int()
    .positive('Parent ID must be a positive integer')
    .optional()
    .describe('Optional parent backlog item ID (for hierarchical items)'),
  
  assigned_to_id: z.number()
    .int()
    .positive('Assigned user ID must be a positive integer')
    .optional()
    .describe('Optional ID of the user assigned to this item'),
  
  label: z.string()
    .max(50, 'Label must be 50 characters or less')
    .optional()
    .describe('Optional label or tag for categorization'),
  
  acceptance_criteria: z.array(z.string().min(1, 'Acceptance criteria cannot be empty'))
    .max(10, 'Maximum 10 acceptance criteria allowed')
    .optional()
    .describe('Optional array of acceptance criteria for the backlog item')
});

/**
 * Schema for the successful backlog creation response
 */
export const backlogCreationResponseSchema = z.object({
  success: z.boolean(),
  backlog_item: z.object({
    id: z.number(),
    title: z.string(),
    description: z.string().optional(),
    priority: z.nativeEnum(BacklogPriority),
    status: z.nativeEnum(BacklogStatus),
    item_type: z.nativeEnum(BacklogType),
    story_point: z.number().optional(),
    project_id: z.number(),
    created_at: z.string(),
    updated_at: z.string()
  }),
  message: z.string()
});

/**
 * Alternative backlog item schema using string enums (for compatibility)
 */
export const backlogItemSchema = z.object({
  title: z.string().min(1).max(200).describe('The title of the backlog item'),
  description: z.string().max(2000).optional().describe('Optional detailed description'),
  priority: z.enum(['critical', 'high', 'medium', 'low']).default('medium').describe('Priority level'),
  status: z.enum(['todo', 'in_progress', 'in_review', 'done', 'cancelled']).default('todo').describe('Current status'),
  item_type: z.enum(['epic', 'story', 'bug']).default('story').describe('Type of backlog item'),
  story_point: z.number().int().min(0).max(100).optional().describe('Estimation in story points'),
  project_id: z.number().int().positive().describe('Project ID'),
  parent_id: z.number().int().positive().optional().describe('Parent backlog item ID'),
  assigned_to_id: z.number().int().positive().optional().describe('Assigned user ID'),
  label: z.string().max(50).optional().describe('Optional label or tag'),
  acceptance_criteria: z.array(z.string().min(1)).max(10).optional().describe('Optional acceptance criteria'),
});

export const backlogRetrievalSchema = z.object({
  project_id: z.number().int().positive().describe('Project ID'),
  status: z.enum(['todo', 'in_progress', 'in_review', 'done', 'cancelled']).optional().describe('Filter by status'),
  priority: z.enum(['critical', 'high', 'medium', 'low']).optional().describe('Filter by priority'),
  item_type: z.enum(['epic', 'story', 'bug']).optional().describe('Filter by type'),
  search: z.string().optional().describe('Search term across fields'),
  label: z.string().optional().describe('Filter by label'),
  assigned_to_id: z.number().int().positive().optional().describe('Filter by assignee'),
  parent_id: z.number().int().positive().optional().describe('Filter by parent item'),
  limit: z.number().int().min(1).max(100).default(50).describe('Max results'),
  skip: z.number().int().min(0).default(0).describe('Offset'),
});

export const updateBacklogItemSchema = backlogItemSchema.partial().extend({
  backlog_id: z.number().int().positive().describe('Backlog item ID to update'),
});

// Export TypeScript types
export type CreateBacklogItemInput = z.infer<typeof createBacklogItemSchema>;
export type BacklogCreationResponse = z.infer<typeof backlogCreationResponseSchema>;
export type BacklogItemInput = z.infer<typeof backlogItemSchema>;
export type BacklogRetrievalInput = z.infer<typeof backlogRetrievalSchema>;
export type UpdateBacklogItemInput = z.infer<typeof updateBacklogItemSchema>;


